In [1]:
from keras.datasets import boston_housing
import numpy

Using TensorFlow backend.
/usr/local/anaconda3/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:526: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/usr/local/anaconda3/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:527: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/usr/local/anaconda3/lib/python3.7/site-packages/tensorflow/python/framework/dtypes.py:528: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/usr/local/anaconda3/lib/python3.7/site-packages/tensorflow/py

In [2]:
(x_train, y_train), (x_test, y_test) = boston_housing.load_data()
print('shape of x_train' + str(x_train.shape))
print('shape of y_train' + str(y_train.shape) + "\n")
print('shape of x_test' + str(x_test.shape))
print('shape of y_test' + str(y_test.shape))


shape of x_train(404, 13)
shape of y_train(404,)

shape of x_test(102, 13)
shape of y_test(102,)


In [3]:
n, d = x_train.shape
xbar_train = numpy.concatenate((x_train, numpy.ones((n,1))), axis=1)

In [4]:
print("shape of x_train:" + str(x_train.shape))
print("shape of xbar_train:" + str(xbar_train.shape))

shape of x_train:(404, 13)
shape of xbar_train:(404, 14)


<img src="./images/normal_equation.png"/>

This equation is the normal equation. It allows us to find a w such that we minimize the cost function 

In [8]:
# red part of equation 
xx = numpy.dot(xbar_train.T, xbar_train)
xx_inv = numpy.linalg.pinv(xx)

In [9]:
#blue part of the equation 
xy = numpy.dot(xbar_train.T, y_train)

In [11]:
# find w 
w = numpy.dot(xx_inv, xy)
print(w)

[-1.19997513e-01  5.70003304e-02  3.98379660e-03  4.12698187e+00
 -2.05002963e+01  3.38024903e+00  7.56807584e-03 -1.71189793e+00
  3.34747537e-01 -1.17797225e-02 -9.02318039e-01  8.71912756e-03
 -5.55842510e-01  4.02936706e+01]


Now we want to plug the w value into our MSE. The reason we're finding MSE is to find out whether our model is good. 

<img src="./images/mse.png"/>

In [14]:
y_lsr = numpy.dot(xbar_train, w)#right hand side of equation
diff = y_lsr - y_train #absolute value, order doesn't matter

MSE = numpy.mean(diff*diff)
print("Train MSE: " + str(MSE))

#so our model is good! 

Train MSE: 22.004800838348142


Remember that w bar is the linear regressor. The label (price of house) is our label or y. The features or x are beds, bad, square feet, cars, etc. 

<img src="./images/overall.png" />

so now, let's test it!

1. Add a feature to test the feature matrix x to xbar 
2. Make prediction by: y_pred = xbar*w 

In [16]:
n_test, _ = x_test.shape 
xbar_test = numpy.concatenate((x_test, numpy.ones((n_test,1))), axis =1)
y_pred = numpy.dot(xbar_test,w)

Now find the MSE of the test 

In [17]:
diff = y_pred - y_test 
MSE = numpy.mean(diff*diff)
print("Test MSE: " + str(MSE))

Test MSE: 23.19559925641708


Test MSE (out of sample error) is 23.2 and Training MSE (in sample error)is 22. 

What do these numbers mean? 

The game is an indicator of generalization 
